crates/polars-ops/src/chunked_array/list/product.rs

```diff
@@ -0,0 +1,181 @@
+            out.into_series()
+        },
+        Float32 => {
+            let out: Float32Chunked =
+                ca.apply_amortized_generic(|s| s.map(|s| product::<f32>(s.as_ref()).unwrap()));
+            out.into_series()
+        },
+        Float64 => {
+            let out: Float64Chunked =
+                ca.apply_amortized_generic(|s| s.map(|s| product::<f64>(s.as_ref()).unwrap()));
+            out.into_series()
+        },
+        _ => panic!("Unsupported data type"),
+    };
+    out.rename(ca.name());
+    Ok(out)
+}
+
+fn product_between_offsets<T, S>(values: &[T], offset: &[i64]) -> Vec<S>
+where
+    T: NativeType + ToPrimitive,
+    S: NumCast + std::iter::Product,
+{
+    let mut running_offset = offset[0];
+
+    (offset[1..])
+        .iter()
+        .map(|end| {
+            let current_offset = running_offset;
+            running_offset = *end;
+
+            let slice = unsafe { values.get_unchecked(current_offset as usize..*end as usize) };
+            product_slice(slice)
+        })
+        .collect_trusted()
+}
+
+fn dispatch_product<T, S>(arr: &dyn Array, offsets: &[i64], validity: Option<&Bitmap>) -> ArrayRef
+where
+    T: NativeType + ToPrimitive,
+    S: NativeType + NumCast + std::iter::Product,
+{
+    let values = arr.as_any().downcast_ref::<PrimitiveArray<T>>().unwrap();
+    let values = values.values().as_slice();
+    Box::new(PrimitiveArray::from_data_default(
+        product_between_offsets::<_, S>(values, offsets).into(),
+        validity.cloned(),
+    )) as ArrayRef
+}
+
+pub(super) fn product_list_numerical(ca: &ListChunked, inner_type: &DataType) -> Series {
+    use DataType::*;
+    let chunks = ca
+        .downcast_iter()
+        .map(|arr| {
+            let offsets = arr.offsets().as_slice();
+            let values = arr.values().as_ref();
+
+            match inner_type {
+                Int8 => dispatch_product::<i8, i64>(values, offsets, arr.validity()),
+                Int16 => dispatch_product::<i16, i64>(values, offsets, arr.validity()),
+                Int32 => dispatch_product::<i32, i64>(values, offsets, arr.validity()),
+                Int64 => dispatch_product::<i64, i64>(values, offsets, arr.validity()),
+                UInt8 => dispatch_product::<u8, i64>(values, offsets, arr.validity()),
+                UInt16 => dispatch_product::<u16, i64>(values, offsets, arr.validity()),
+                UInt32 => dispatch_product::<u32, i64>(values, offsets, arr.validity()),
+                UInt64 => dispatch_product::<u64, u64>(values, offsets, arr.validity()),
+                Float32 => dispatch_product::<f32, f32>(values, offsets, arr.validity()),
+                Float64 => dispatch_product::<f64, f64>(values, offsets, arr.validity()),
+                _ => unimplemented!(),
+            }
+        })
+        .collect::<Vec<_>>();
+
+    Series::try_from((ca.name(), chunks)).unwrap()
+}
+
+pub(super) fn product_boolean(ca: &ListChunked) -> PolarsResult<Series> {
```

Comment:
I believe this current implementation does not correctly take into account the validity of the inner arrays?

Reply:
Thanks for reviewing this PR.
Does it need to take it into account? If so, how could it use it?
Note that this implementation was heavily based on https://github.com/pola-rs/polars/blob/main/crates/polars-ops/src/chunked_array/list/count.rs#L8